backend/connect.ts

```diff
@@ -0,0 +1,81 @@
+import { IncomingMessage } from "http";
+import { cookieSchema } from "../schemas/poke";
+import { parse } from "url";
+import { WebSocket } from "ws";
+import { getClientRecord, setClientRecord } from "./data";
+import { transact } from "./db";
+import { dispatchMessage } from "./dispatch-message";
+import { Client, ClientMap } from "./server";
+import { Loop } from "./loop";
+
+export async function connect(
+  ws: WebSocket,
+  req: IncomingMessage,
+  clients: ClientMap,
+  loop: Loop
+) {
+  const url = parse(req.url!, true);
+  const parts = (url.pathname ?? "").split("/");
+
+  if (parts[1] != "d" || !parts[2]) {
+    ws.send("invalid url - no room id");
+    ws.close();
+    return;
+  }
+
+  const roomID = parts[2];
+  const clientID = (url.query.clientID as string | undefined) ?? "";
+  if (!clientID) {
+    ws.send("invalid url - no client ID");
+    ws.close();
+    return;
+  }
+  const baseCookie = cookieSchema.safeParse(
+    JSON.parse((url.query.baseCookie as string | undefined) ?? "null")
+  );
+  if (!baseCookie.success) {
+    ws.send(`invalid base cookie in url: ${baseCookie.error.issues}`);
+    ws.close();
+    return;
+  }
+
+  const existingClient = clients.get(clientID);
+  if (existingClient) {
+    existingClient.socket.close();
+  }
+
+  await transact(async (tx) => {
+    let record = await getClientRecord(tx, clientID);
+    if (record) {
+      if (record.documentID != roomID) {
+        ws.send(`Cannot change roomID - already in room ${record.documentID}`);
+        ws.close();
+      }
+      record.baseCookie = baseCookie.data;
+    } else {
+      record = {
+        id: clientID,
+        lastMutationID: 0,
+        baseCookie: baseCookie.data,
+        documentID: roomID,
+      };
+    }
+    await setClientRecord(tx, record);
+  });
+
+  const client: Client = {
+    clientID,
+    roomID,
+    socket: ws,
+    pending: [],
+  };
+  clients.set(clientID, client);
```

Comment:
I don't understand what it would mean to update it transactionally -- if I updated the in-memory state inside the tx then (a) commit might fail and I'd have to roll back? (b) concurrent code could observe the in-memory change before the persistent change?

It seems we have no choice but to document carefully the relationship between the in-memory state and the persistent state. Setting the in-memory change after the persistent change makes sense to me since it cannot fail.